### PR TITLE
Sph efficiency

### PR DESCRIPTION
**Pull request by matthiasto**
Added a cython version of the code.
First compile the .pyx file following the instructions in the file
Choose the options in com1DFA.ini and com1DFA.py

### CODE REVIEW

**Comment by fso42** [comment]
Is this meant as dropin for the pure python version i.e. replacing, or live alongside ?

File: avaframe/avaframeCfg.ini [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 [MAIN]
 # Path to avalanche directory
-avalancheDir = data/avaHockeySmoothSmall
+avalancheDir = data/avaInclinedPlane
```
Comment (fso42) [inline avaframe/avaframeCfg.ini:7]: I think we should keep the Hockey as default

File: avaframe/com1DFAPy/SPHfunctions.py [hunk 2]
```diff
@@ -139,11 +146,13 @@ def getNeighboursVect(particles, dem):
 
     # make the list of which particles are in which cell
     partInCell = np.argsort(ic, kind='mergesort')
-    InCell = np.vstack((indx, indy))
-    InCell = np.vstack((InCell, ic))
-    InCell = InCell.T
+    # InCell = np.vstack((indx, indy))
+    # InCell = np.vstack((InCell, ic))
+    # InCell = InCell.T
```
Comment (fso42) [inline avaframe/com1DFAPy/SPHfunctions.py:151]: remove this?

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx [hunk 3]
```diff
@@ -0,0 +1,1264 @@
...
+# 2) project on the plane, compute the gradient in the local coord sys related
+# to the local plane and flow direction (orthogonal coord sys)
+# enables to choose earth pressure coefficients
+# 3) project on the plane, compute the gradient in the local coord sys related
+# to the local plane (tau1, tau2, n) non orthogonal coord sys
+cdef int SPHOption = 2
+
+cdef double rho = 200
+cdef double csz = 5
+cdef double gravAcc = 9.81
```
Comment (fso42) [inline avaframe/com1DFAPy/SPHfunctionsCython.pyx:40]: These need to either be moved out to the ini right away, or mark them for moving with a big todo or development flag

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx [hunk 4]
```diff
@@ -0,0 +1,1264 @@
...
+  log.debug('dt used now is %f' % DT)
+  cdef double gravAcc = cfg.getfloat('gravAcc')
+  cdef double rho = cfg.getfloat('rho')
+  cdef double csz = dem['header'].cellsize
+  cdef double mu = cfg.getfloat('mu')
+  cdef int Npart = particles['Npart']
+  cdef double[:, :] Nx = dem['Nx']
+  cdef double[:, :] Ny = dem['Ny']
+  cdef double[:, :] Nz = dem['Nz']
+  cdef double[:, :] Z = dem['rasterData']
```
Comment (fso42) [inline avaframe/com1DFAPy/SPHfunctionsCython.pyx:309]: Please fix the indentation

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx [hunk 5]
```diff
@@ -0,0 +1,1264 @@
...
+    hBB[j] = hbb
+  # hB, _ = geoTrans.projectOnRasterVectRoot(particles['x'], particles['y'], fields['FD'], csz=csz, interp='bilinear')
+  # indxx = particles['indX']
+  # indyy = particles['indY']
+  # # APart = dem['Area'][indyy, indxx]
+  # hLim = 400*25/cfg.getfloat('rho') #particles['m']/APart/cfg.getfloat('rho')
+  # hB = np.where(hB < hLim, hLim, hB)
+  # particles['hBilinearBilinear'] = hB
+  # particles['h'] = hB
+# choose the interpolation method
```
Comment (fso42) [inline avaframe/com1DFAPy/SPHfunctionsCython.pyx:574]: Still needed?

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx [hunk 6]
```diff
@@ -0,0 +1,1264 @@
...
+cdef double csz = 5
+cdef double gravAcc = 9.81
+
+ctypedef double dtypef_t
+ctypedef long dtypel_t
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def pointsToRasterC(x, y, z, Z0, csz=1, xllc=0, yllc=0):
```
Comment (fso42) [inline avaframe/com1DFAPy/SPHfunctionsCython.pyx:48]: Is the C in all these names necessary? -> relates to my question whether it is a replacement or duplicate...

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx [hunk 7]
```diff
@@ -0,0 +1,1264 @@
...
+                #           dwdr = dfacKernel * hr * hr
+                #           mdwdrr = mass[l] * dwdr / r
+                #           gradhX = gradhX + mdwdrr*dx
+                #           gradhY = gradhY + mdwdrr*dy
+                #
+                # if SPHoption == 2:
+                # get coordinates in local coord system
+                r1 = scalProd(dx, dy, dz, ux, uy, uz)
+                r2 = scalProd(dx, dy, dz, uxOrtho, uyOrtho, uzOrtho)
+                # impse r3=0 even if the particle is not exactly on the tengent plane
```
Comment (fso42) [inline avaframe/com1DFAPy/SPHfunctionsCython.pyx:863]: Feature flag and extra functions?

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx [hunk 8]
```diff
@@ -0,0 +1,1264 @@
...
+  Lx0 = <int>Lx
+  Ly0 = <int>Ly
+  Lx1 = Lx0 + 1
+  Ly1 = Ly0 + 1
+  # prepare for bilinear interpolation
+  dx = Lx - Lx0
+  dy = Ly - Ly0
+  # using bilinear interpolation on the cell
+
+  cdef double v = V[Ly0, Lx0]*(1-dx)*(1-dy) + V[Ly0, Lx1]*dx*(1-dy) + V[Ly1, Lx0]*(1-dx)*dy + V[Ly1, Lx1]*dx*dy
```
Comment (fso42) [inline avaframe/com1DFAPy/SPHfunctionsCython.pyx:1194]: make this multiline

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx [hunk 9]
```diff
@@ -0,0 +1,1264 @@
...
+  Ly1 = Ly0 + 1
+  # prepare for bilinear interpolation
+  dx = Lx - Lx0
+  dy = Ly - Ly0
+  # using bilinear interpolation on the cell
+
+  cdef double nx = Nx[Ly0, Lx0]*(1-dx)*(1-dy) + Nx[Ly0, Lx1]*dx*(1-dy) + Nx[Ly1, Lx0]*(1-dx)*dy + Nx[Ly1, Lx1]*dx*dy
+  cdef double ny = Ny[Ly0, Lx0]*(1-dx)*(1-dy) + Ny[Ly0, Lx1]*dx*(1-dy) + Ny[Ly1, Lx0]*(1-dx)*dy + Ny[Ly1, Lx1]*dx*dy
+  cdef double nz = Nz[Ly0, Lx0]*(1-dx)*(1-dy) + Nz[Ly0, Lx1]*dx*(1-dy) + Nz[Ly1, Lx0]*(1-dx)*dy + Nz[Ly1, Lx1]*dx*dy
+  return nx, ny, nz
```
Comment (fso42) [inline avaframe/com1DFAPy/SPHfunctionsCython.pyx:1256]: multiline for readability. 
Another note: did you try to use one of numpys (or similar) interpolation routines?

File: avaframe/com1DFAPy/Tcpu.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+from avaframe.out3Plot.plotUtils import *
+import avaframe.com1DFAPy.timeDiscretizations as tD
+import avaframe.com1DFAPy.DFAtools as DFAtls
+import avaframe.com1DFAPy.SPHfunctions as SPH
+import avaframe.com1DFAPy.frictionLaws as fricLaws
+
+from SPHfunctionsCython import *
+# import avaframe.in2Trans.shpConversion as shpConv
+# import avaframe.in2Trans.ascUtils as IOf
+# from avaframe.DFAkernel.setParam import *
```
Comment (fso42) [inline avaframe/com1DFAPy/Tcpu.py:25]: remove?

File: avaframe/com1DFAPy/Tcpu.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+# create local logger
+log = logging.getLogger(__name__)
+
+slopeAnglex = 40*math.pi/180
+slopeAngley = 20*math.pi/180
```
Comment (fso42) [inline avaframe/com1DFAPy/Tcpu.py:31]: What is this for?

File: avaframe/com1DFAPy/Tcpu.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
+# import pyximport
+# pyximport.install()
+
+import logging
+import time
```
Comment (fso42) [inline avaframe/com1DFAPy/Tcpu.py:5]: Maybe a quick note what this module is used for?

File: avaframe/runCom1DFA.py [hunk 13]
```diff
@@ -33,7 +33,7 @@
 startTime = time.time()
 
 # Clean input directory(ies) of old work and output files
-initProj.cleanSingleAvaDir(avalancheDir, keep=logName)
+initProj.cleanSingleAvaDir(avalancheDir, keep=logName, deleteOutput=False)
```
Comment (fso42) [inline avaframe/runCom1DFA.py:36]: I think we should leave the default option of deleting the output folder as well...

**Review by fso42 — CHANGES_REQUESTED** [review]
A general note, but not necessary for this PR: break out logical parts into smaller functions, so that the main reads more like an algorithm and it makes it easier to switch out different methods...

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx (hunk 6 above)
Comment (matthiasto) [reply avaframe/com1DFAPy/SPHfunctionsCython.pyx:48]: You dont need the C in the name. This function does the same as pointtoRaster in geotrans but purely in C.
It is the case for almost all the NameOfFunctionC(), it is the C version of NameOfFunction. By choosing the flag flagCython you can decide to use cython functions or numpy ones

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx (hunk 8 above)
Comment (matthiasto) [reply avaframe/com1DFAPy/SPHfunctionsCython.pyx:1194]: you mean make intermediate variables or just split the line?

File: avaframe/com1DFAPy/SPHfunctionsCython.pyx (hunk 9 above)
Comment (matthiasto) [reply avaframe/com1DFAPy/SPHfunctionsCython.pyx:1256]: the interpolation methods in scipy did not seem faster. (there is already the numpy version of this cython function which is I think slower)

File: avaframe/com1DFAPy/Tcpu.py (hunk 11 above)
Comment (matthiasto) [reply avaframe/com1DFAPy/Tcpu.py:31]: I will remove those functions from the git repository

**Pull request merged**